### PR TITLE
[Validator] reset the validation context after validating nested constraints

### LINKED ISSUE #54577: Validation skips violations in case with nested AtLeastOneOf constraint
### Symfony version(s) affected

6.4

### Description

When I make nested constraints **AtLeastOneOf**, the validator will skip all violations.

The Symfony validator employs two factories for initializing an instance of a constraint validator:
1.  Symfony\Component\ValidatorConstraintValidatorFactory - creates a new instance of the validator and caches it as an array.
2. Symfony\Component\ContainerConstraintValidatorFactory - retrieves the validator from the DI container. As each unique service in a container is passed by reference, each validator is initialized only once.

The issue arises when creating a nested constraint, such as `AtLeastOneOf`. At line 752 of `RecursiveContextualValidator`, a new context for this validator is initialized, which is the root cause of the bug.

### How to reproduce

To reproduce this bug, you can use the following code:

```php
<?php

declare(strict_types=1);

use Symfony\Component\Validator\Constraints\AtLeastOneOf;
use Symfony\Component\Validator\Constraints\Choice;
use Symfony\Component\Validator\Constraints\Collection;
use Symfony\Component\Validator\Constraints\Type;
use Symfony\Component\Validator\Constraints\NotBlank;
use Symfony\Component\Validator\Validation;

require_once dirname(__DIR__) . '/vendor/autoload.php';

$values = [
    'AtLeastOneOf' => [
        'field1' => 'test',
        'AtLeastOneOfNested1' => [
            'AtLeastOneOfNested2' => \uniqid(),
            'field2' => \uniqid(),
        ],
    ],
    'test2' => '1',
];

$constraints = new Collection([
    'AtLeastOneOf' => new AtLeastOneOf([
        new Collection([
            'field1' => new NotBlank(),
            'AtLeastOneOfNested1' => new AtLeastOneOf([
                new Collection([
                    'AtLeastOneOfNested2' => new AtLeastOneOf([
                        new Type('int'),
                        new Choice(['test1', 'test2'])
                    ]),
                ]),
                new Collection([
                    'field2' => new Type('int'),
                ]),
            ]),
        ]),
        new Collection([
            'field3' => new NotBlank(),
            'AtLeastOneOfNested3' => new AtLeastOneOf([
                new Collection([
                    'field1' => new Type('int'),
                ]),
                new Collection([
                    'field2' => new Type('int'),
                ]),
            ]),
        ]),
    ]),
    'test2' => new NotBlank(),
]);

$violations = Validation::createValidator()->validate($values, $constraints);

echo $violations->count() . \PHP_EOL;
```

As we can see, incorrect data exists for the key `AtLeastOneOfNested1`, but the validator skips these violations.

### Possible Solution

A possible solution is to consistently initialize a new instance of `AtLeastOneOfValidator` within a factory.

image
image

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by xabbuh**
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Issues        | Fix #54577
| License       | MIT

### CODE REVIEW

**Comment by xabbuh** [comment]
I still need to push a test covering the bug described in #54577, wanted to confirm first that the change doesn't break anything

Status: Needs work

**Comment by xabbuh** [comment]
test added

Status: Needs Review

File: src/Symfony/Component/Validator/Constraints/AtLeastOneOfValidator.php [hunk 1]
```diff
@@ -42,9 +42,11 @@ public function validate($value, Constraint $constraint)
                 continue;
             }
 
+            $context = $this->context;
             $executionContext = clone $this->context;
             $executionContext->setNode($value, $this->context->getObject(), $this->context->getMetadata(), $this->context->getPropertyPath());
             $violations = $validator->inContext($executionContext)->validate($value, $item, $this->context->getGroup())->getViolations();
+            $this->context = $context;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/Constraints/AtLeastOneOfValidator.php:49]: just wondering: would it make sense to nest this in a finally block?

Comment (xabbuh) [reply src/Symfony/Component/Validator/Constraints/AtLeastOneOfValidator.php:49]: I don't think so. First, we don't do that anywhere else and second `validate()` is not supposed to throw. If it did, I would expect more things to no longer work as expected.

**Review by nicolas-grekas — APPROVED** [review]

**Review by OskarStark — APPROVED** [review]

**Pull request merged**